File: ios/bridge/ExpoBridgeRatings.swift
import Foundation

@available(iOS 16.0, *)
enum ExpoBridgeRatings {
  static func getRating(
    service: RatingsService,
    musicUserToken: String,
    resourceType: String,
    id: String
  ) async throws -> [String: Any]? {
    try await AppleMusicBridgeError.rethrow {
      try await service.getRating(musicUserToken: musicUserToken, resourceType: resourceType, id: id)
    }
  }

  static func setRating(
    service: RatingsService,
    musicUserToken: String,
    resourceType: String,
    id: String,
    value: Int
  ) async throws -> [String: Any] {
    try await AppleMusicBridgeError.rethrow {
      try await service.setRating(
        musicUserToken: musicUserToken, resourceType: resourceType, id: id, value: value)
    }
  }

  static func clearRating(
    service: RatingsService,
    musicUserToken: String,
    resourceType: String,
    id: String
  ) async throws {
    try await AppleMusicBridgeError.rethrow {
      try await service.clearRating(
        musicUserToken: musicUserToken, resourceType: resourceType, id: id)
    }
  }

  static func addToFavorites(
    service: RatingsService,
    musicUserToken: String,
    resourceIds: [String: [String]]
  ) async throws {
    try await AppleMusicBridgeError.rethrow {
      try await service.addToFavorites(musicUserToken: musicUserToken, resourceIds: resourceIds)
    }
  }

  static func removeFromFavorites(
    service: RatingsService,
    musicUserToken: String,
    resourceIds: [String: [String]]
  ) async throws {
    try await AppleMusicBridgeError.rethrow {
      try await service.removeFromFavorites(musicUserToken: musicUserToken, resourceIds: resourceIds)
    }
  }
}

File: android/src/main/java/expo/modules/applemusic/bridge/ExpoBridgeRatings.kt
package expo.modules.applemusic.bridge

import expo.modules.applemusic.RatingsRestClient
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.ModuleDefinitionBuilder

internal fun ModuleDefinitionBuilder.registerRatingsBridge(
  ratings: () -> RatingsRestClient,
) {
  AsyncFunction("getRating") Coroutine { musicUserToken: String, resourceType: String, id: String ->
    ratings().getRating(musicUserToken, resourceType, id)
  }

  AsyncFunction("setRating") Coroutine { musicUserToken: String, resourceType: String, id: String, value: Int ->
    ratings().setRating(musicUserToken, resourceType, id, value)
  }

  AsyncFunction("clearRating") Coroutine { musicUserToken: String, resourceType: String, id: String ->
    ratings().clearRating(musicUserToken, resourceType, id)
  }

  AsyncFunction("addToFavorites") Coroutine { musicUserToken: String, resourceIds: Map<String, List<String>> ->
    ratings().addToFavorites(musicUserToken, resourceIds)
  }

  AsyncFunction("removeFromFavorites") Coroutine { musicUserToken: String, resourceIds: Map<String, List<String>> ->
    ratings().removeFromFavorites(musicUserToken, resourceIds)
  }
}
